ENH: Add support for scikit-sparse v0.5.0

broesler opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### What does this implement/fix?
This PR adds support for [scikit-sparse v0.5.0](https://github.com/scikit-sparse/scikit-sparse), namely, the `cholmod` and `umfpack` submodules. The `cholmod` module in scikit-sparse v0.4.16 has been largely rewritten to remove bugs and run more efficiently with a Cython interface. The scikit-sparse `umfpack` submodule is new as of v0.5.0, and replaces the [scikits-umfpack](https://github.com/scikit-umfpack/scikit-umfpack) project, which is largely dormant. See [this SciPy forum discussion](https://discuss.scientific-python.org/t/improve-interface-to-suitesparse-umfpack/2138) for more details.

The only external API change in SciPy is that the `sparse.linalg.use_solver` function has been removed along with the global variable `useUmfpack` used in `sparse.linalg._dsolve`. The `use_umfpack` kwarg alone controls the use of UMFPACK. This PR keeps the existing behavior of using the UMFPACK factorization/solver if scikit-sparse is installed, but silently defaulting to SuperLU factorization/solver otherwise.

Additional changes:
* `sparse.linalg.spsolve` raises `LinAlgError` for a singular matrix, regardless of the RHS input type (sparse vs. dense, 1D vs. 2D). Previously, a dense RHS would *not* raise, but return a vector of `np.nan`, whereas a sparse RHS would raise `RuntimeError` when `use_umfpack=False`.
* The `sparse.linalg.MatrixRankWarning` class has been removed.


#### Additional information
This PR will remain a draft until the scikit-sparse v0.5.0 changes are all merged into master. The development branch is available [on my fork](https://github.com/broesler/scikit-sparse).

broesler committed:
MNT: Update sksparse.cholmod usage to v0.5.0 API.

broesler committed:
MNT: Update sksparse.cholmod usage to v0.5.0 API.

broesler committed:
TST: Update sksparse.cholmod usage in tests to v0.5.0 API.

broesler committed:
TST: Update sksparse.cholmod usage in tests to v0.5.0 API.

broesler committed:
MNT: Replace scikits.umfpack with sksparse.umfpack (v0.5.0) linprog_ip.

broesler committed:
MNT: Replace scikits.umfpack with sksparse.umfpack (v0.5.0) linprog_ip.

broesler committed:
TST: Replace scikits.umfpack with sksparse.umfpack (v0.5.0) in tests.

This commit updates test_linprog.py to use the new API of sksparse
v0.5.0. It removes many unnecessary warning filters.

broesler committed:
TST: Replace scikits.umfpack with sksparse.umfpack (v0.5.0) in tests.

This commit updates test_linprog.py to use the new API of sksparse
v0.5.0. It removes many unnecessary warning filters.

broesler committed:
ENH: Replace scikit.umfpack with sksparse.umfpack in linsolve.

This commit replaces all uses of scikit.umfpack with the
sksparse.umfpack v0.5.0 API.

It also incorporates changes from feature/spsolve-shape and
feature/spsolve-batches in anticipation of their merge into main before
this PR is reviewed.

broesler committed:
ENH: Replace scikit.umfpack with sksparse.umfpack in linsolve.

This commit replaces all uses of scikit.umfpack with the
sksparse.umfpack v0.5.0 API.

It also incorporates changes from feature/spsolve-shape and
feature/spsolve-batches in anticipation of their merge into main before
this PR is reviewed.

broesler committed:
TST: Replace scikit.umfpack with sksparse.umfpack in test_linsolve.

This commit updates test_linsolve.py to use sksparse.umfpack, and
updates some warnings and other messages to be compatible.

This commit incorporates changes from feature/spsolve-shape in
anticipation of that commit being merged into main before this PR is
created.

broesler committed:
TST: Replace scikit.umfpack with sksparse.umfpack in test_linsolve.

This commit updates test_linsolve.py to use sksparse.umfpack, and
updates some warnings and other messages to be compatible.

This commit incorporates changes from feature/spsolve-shape in
anticipation of that commit being merged into main before this PR is
created.

broesler committed:
ENH: Remove global variable useUmfpack from _dsolve/linsolve.py.

This commit removes the use of a global variable for solver
configuration, and requires the kwarg "use_umfpack=True/False" where
appropriate. The existing behavior of "use umfpack if it is installed,
otherwise fall back silently to SuperLU" is retained.

broesler committed:
ENH: Remove global variable useUmfpack from _dsolve/linsolve.py.

This commit removes the use of a global variable for solver
configuration, and requires the kwarg "use_umfpack=True/False" where
appropriate. The existing behavior of "use umfpack if it is installed,
otherwise fall back silently to SuperLU" is retained.

broesler committed:
TST: Update test_linsolve with explicity use_umfpack usage.

This commit removes the use_solver function calls in favor of the
explicit use_umfpack kwarg. In some test classes, the setup method
defines self.spsolve as a partial function with use_umfpack=True/False
instead of setting a global variable within the class.

broesler committed:
TST: Update test_linsolve with explicity use_umfpack usage.

This commit removes the use_solver function calls in favor of the
explicit use_umfpack kwarg. In some test classes, the setup method
defines self.spsolve as a partial function with use_umfpack=True/False
instead of setting a global variable within the class.

broesler committed:
ENH: Raise LinAlgError for singular matrix in spsolve.

This commit changes the behavior of spsolve with a dense `b` input.
Previously, it would warn with MatrixRankWarning and return a vector of
np.nan values. For a sparse `b` input, spsolve would raise. Now, it
raises on either sparse or dense input.

This commit removes the MatrixRankWarning class, as it is no longer
needed.

This commit changes the type of error raised for a sparse `b` input.
SuperLU raises a RuntimeError when factorizing a singular matrix.
Previously, spsolve just raised that error. Now, it raises a LinAlgError
from that error.

The error behavior for a singular matrix is now consistent across all
input types and use_umfpack=True/False.

broesler committed:
ENH: Raise LinAlgError for singular matrix in spsolve.

This commit changes the behavior of spsolve with a dense `b` input.
Previously, it would warn with MatrixRankWarning and return a vector of
np.nan values. For a sparse `b` input, spsolve would raise. Now, it
raises on either sparse or dense input.

This commit removes the MatrixRankWarning class, as it is no longer
needed.

This commit changes the type of error raised for a sparse `b` input.
SuperLU raises a RuntimeError when factorizing a singular matrix.
Previously, spsolve just raised that error. Now, it raises a LinAlgError
from that error.

The error behavior for a singular matrix is now consistent across all
input types and use_umfpack=True/False.

broesler committed:
TST: Update test_linsolve to handle LinAlgError for singular matrices.

broesler committed:
TST: Update test_linsolve to handle LinAlgError for singular matrices.

broesler committed:
TST: Remove MatrixRankWarning from test_linprog.

broesler committed:
TST: Remove MatrixRankWarning from test_linprog.

broesler committed:
ENH: Implement `trans` kwarg in sparse.linalg.spsolve.

This commit adds the existing functionality of SuperLU.solve(...,
trans='N'), and sksparse.umfpack.umf_solve(..., trans='N'), etc. to the
higher-level spsolve function.

The previously removed _superlu.gssv does not support the "Trans" option
(it can be passed in, but does nothing).

broesler committed:
ENH: Implement `trans` kwarg in sparse.linalg.spsolve.

This commit adds the existing functionality of SuperLU.solve(...,
trans='N'), and sksparse.umfpack.umf_solve(..., trans='N'), etc. to the
higher-level spsolve function.

The previously removed _superlu.gssv does not support the "Trans" option
(it can be passed in, but does nothing).

broesler committed:
ENH: Update optimize SuperLU.solve call to general spsolve.

Refactor:
    `splu(B).solve(e, trans='T')` -> `spsolve(B, e, trans='T')`

Since sksparse.umfpack.umf_solve can handle the `trans` kwarg, use the
general spsolve function instead of restricting to the SuperLU call.

broesler committed:
ENH: Update optimize SuperLU.solve call to general spsolve.

Refactor:
    `splu(B).solve(e, trans='T')` -> `spsolve(B, e, trans='T')`

Since sksparse.umfpack.umf_solve can handle the `trans` kwarg, use the
general spsolve function instead of restricting to the SuperLU call.

broesler committed:
TST: Add tests for spsolve(..., trans='T') etc. to test_linsolve.py.

broesler committed:
TST: Add tests for spsolve(..., trans='T') etc. to test_linsolve.py.

broesler committed:
ENH: Remove _superlu.gssv call and only allow CSC format in spsolve.

This commit is the result of performance testing showing that gssv calls
are equivalent to splu(A).solve(b) for CSC input, but substantially
slower for large matrices and many RHS for CSR input. It is faster to
just convert the input to CSC format and call splu(A).solve(b) for all
but small matrices (N ~ 400) and/or small numbers of RHS (K < 10).

broesler committed:
ENH: Remove _superlu.gssv call and only allow CSC format in spsolve.

This commit is the result of performance testing showing that gssv calls
are equivalent to splu(A).solve(b) for CSC input, but substantially
slower for large matrices and many RHS for CSR input. It is faster to
just convert the input to CSC format and call splu(A).solve(b) for all
but small matrices (N ~ 400) and/or small numbers of RHS (K < 10).

broesler committed:
TST: Update test suite to handle spsolve CSC requirement.

This commit updates some tests to keep spsolve from issuing
a SparseEfficiencyWarning when given a CSR format input. The commit
makes two types of changes:
  1. An additional warning filter is added, or expected warning message
     changed.
  2. Test inputs that are arbitrarily CSR format are changed to CSC.

Some tests still fail in these files that require further changes to the
underlying code that calls spsolve:
  * scipy/interpolate/tests/test_bsplines.py
  * scipy/optimize/tests/test__remove_redundancy.py
  * scipy/optimize/tests/test_linprog.py

broesler committed:
TST: Update test suite to handle spsolve CSC requirement.

This commit updates some tests to keep spsolve from issuing
a SparseEfficiencyWarning when given a CSR format input. The commit
makes two types of changes:
  1. An additional warning filter is added, or expected warning message
     changed.
  2. Test inputs that are arbitrarily CSR format are changed to CSC.

Some tests still fail in these files that require further changes to the
underlying code that calls spsolve:
  * scipy/interpolate/tests/test_bsplines.py
  * scipy/optimize/tests/test__remove_redundancy.py
  * scipy/optimize/tests/test_linprog.py

broesler committed:
MNT: Convert bspline design matrix to CSC format for spsolve.

This commit is in response to the removal of the _superlu.gssv call in
sparse.linalg.spsolve, which requires spsolve to only accept CSC inputs.

broesler committed:
MNT: Convert bspline design matrix to CSC format for spsolve.

This commit is in response to the removal of the _superlu.gssv call in
sparse.linalg.spsolve, which requires spsolve to only accept CSC inputs.

broesler committed:
ENH: Refactor spsolve(B.T, e) -> splu(B).solve(e, trans='T').

This commit is in response to the removal of the _superlu.gssv call in
sparse.linalg.spsolve, which requires spsolve to only accept CSC inputs.

The B matrix is in CSC format, so its transpose is in CSR format, which
is not supported by spsolve. It is more efficient to create the
SuperLU factorization of the CSC format, then call the solve method with
`trans='T'` to handle the "transpose" internally at the C level.

This change forces use of SuperLU instead of the more generic spsolve,
but a future commit will support the `trans` argument to spsolve.

broesler committed:
ENH: Refactor spsolve(B.T, e) -> splu(B).solve(e, trans='T').

This commit is in response to the removal of the _superlu.gssv call in
sparse.linalg.spsolve, which requires spsolve to only accept CSC inputs.

The B matrix is in CSC format, so its transpose is in CSR format, which
is not supported by spsolve. It is more efficient to create the
SuperLU factorization of the CSC format, then call the solve method with
`trans='T'` to handle the "transpose" internally at the C level.

This change forces use of SuperLU instead of the more generic spsolve,
but a future commit will support the `trans` argument to spsolve.

broesler committed:
DOC: Update scipy.sparse spsolve doc to use CSC format.

broesler committed:
DOC: Update scipy.sparse spsolve doc to use CSC format.

broesler committed:
DOC: Update rhs_batch_size doc to include sksparse.umfpack.

broesler committed:
DOC: Update rhs_batch_size doc to include sksparse.umfpack.

broesler commented:
# Points of Discussion
The new `sksparse.umfpack` module only accepts CSC format matrices, which is
consistent with SuperLU (via `scipy.sparse.linalg.splu`). Non-CSC inputs are
converted to CSC and issue a `SparseEfficiencyWarning`. The existing
`scikits.umfpack` module accepts CSR format matrices without warning, but its
implementation is buggy:

* The factorization routine just converts to CSC internally with `mtx.tocsc()`
  and *recomputes* the factorization to return L, U, P, Q, R.
* If a `scikits.umfpack.UmfpackLU` object is created with a CSR input, then any
  of the L, U, P, Q, R matrices are accessed, the object's `solve` method will
  raise a `ValueError` because the internal factorization has been converted to
  CSC format independently of the input matrix CSR format.
* The solve routine uses `autoTranspose=True` to switch the system from
  A x = b to A.T x = b, but raises a `RuntimeError` if the input is CSR and the
  system requested is A.T.conj() (*i.e.* `trans='H'` in SuperLU terms), because
  the CSR format is a non-conjugate transpose.

It is possible to implement support for CSR format in `sksparse.umfpack`, but it
creates 2x the code paths for most operations, which puts a large burden on
maintenance and testing. It also means that the row scaling matrix output
becomes a *column* scaling matrix if the input is in CSR format, which would
likely be unexpected and confusing to the user.

## Performance Testing of SuperLU with CSR input
The consequence of this behavior is that `spsolve` only accepts CSR inputs
without warning when the RHS is dense and scikit-umfpack is not installed. In
this case, the existing `spsolve` directly calls the `_superlu.gssv` function
with an appropriate flag to indicate that the input is in CSR format. I was
curious about this code path, so I did some performance testing to see how it
compares to using `splu(A).solve(b)`. 

### CSC Input Matrix

<img width="1920" height="2400" alt="superlu_solve_perf_CSC" src="https://github.com/user-attachments/assets/d2426fe7-d4cf-4437-9892-538722a6e52f" />

* Minor peak memory increase when using `splu` with `b` shape (N, 1) since
  a temporary `SuperLU` object is created.
* `gssv` is faster for matrices up to size ~ 30 x 30, but runtime and memory use
  are essentially identical for larger matrices.

### CSR Input Matrix

<img width="1920" height="2400" alt="superlu_solve_perf_CSR" src="https://github.com/user-attachments/assets/0eb4cf8b-be7e-4854-aad5-45da0f17df6a" />

* Essentially no difference in peak memory use for any scenario.
* `gssv` is substantially slower than `splu` for all but the smallest matrices.
  The difference becomes more pronounced as the number of RHS vectors increases.
  
[Full testing script](https://github.com/broesler/scipy_dev_scripts/blob/main/superlu_solve_perf.py).

## Conclusions
I have chosen to optimize for large matrices with multiple RHS
vectors, since that is where sparse solvers are most useful. Therefore, I
have removed the `_superlu.gssv` call from `spsolve`, and replaced it with
`splu(A).solve(b)`. This means that `spsolve` now only accepts CSC format inputs,
regardless of whether UMFPACK or SuperLU is used as the underlying solver.

## Downstream Changes
This change led to some minor changes in tests to filter the
`SparseEfficiencyWarning` where CSR inputs were used, or to change the inputs to
CSC format where appropriate.

There are two changes to other parts of the code base as a result:
1. In `scipy/interpolate/_ndbspline.py`, I have added a check if the solver is
   `spsolve`, and if so, convert the input matrix to CSC format before calling
   the solver.
2. In `scipy/optimize/_remove_redundancy.py`, I have replaced a call to
   `spsolve(B.transpose(), e)` with `spsolve(B, e, trans='T')`.
   Since both `splu(A).solve` and `umfpack.UMFFactor.solve` accept a `trans`
   argument, we can now accept `trans` directly to `spsolve`. In this particular 
   use case, `B` is created in CSC format for column slicing, so its transpose
   is in CSR format. By passing `trans='T'`, we can avoid the explicit transpose
   and conversion to CSC format before calling `spsolve`.

**Note**: `_superlu.gssv` accepts an [options dict](https://portal.nersc.gov/project/sparse/superlu/superlu_code_html/dgssv_8c.html) which may contain a `Trans` option,
but it does not seem like the function
respects that option. SciPy's wrapper seems to pass that option through to
SuperLU, but the SuperLU library itself does not seem to support that option in
`gssv`.

These changes fix all of the `SparseEfficiencyWarning` issues in the SciPy test
suite (with or without `sksparse.umfpack` installed). I am open to any feedback
on either the behavioral or implementation changes.

tylerjereddy commented:
This PR has no review comments from SciPy core developers despite being open for almost six months.

It looks like there is support for this, but perhaps there hasn't been enough recent activity to make it safe to merge this just days before branching for the next release.

Perhaps I'll bump the milestone so it doesn't fall off the radar, apologies it has taken a while for us to get to reviewing this.

broesler commented:
Thanks, @tylerjereddy. I am cleaning up some things in scikit-sparse v0.5.0 and can then get this PR updated and convert to being ready for review.

lucascolley commented:
hey @broesler, just wondering if there is any update here?

broesler commented:
I am waiting on review for the [conda-forge PR](https://github.com/conda-forge/scikit-sparse-feedstock/pull/55) of Scikit-Sparse v0.5.0, and then will convert this to a reviewable PR.

broesler committed:
TST: Fix tril, triu imports.

broesler committed:
MNT: Rename "batch" to "block".

The term "batch" is used to N-dimensional batching (beyond 2D), so we
use "block" here to distinguish.

broesler commented:
This PR incorporates changes from #24055 and #24037. Based on the discussion from #24037, I am fine with reverting to the old behavior regarding size (N, 1) RHS, but the current behavior in this PR is consistent with the SuperLU and Scikit-Sparse packages.

lucascolley commented:
> The only external API change in SciPy is that the `sparse.linalg.use_solver` function has been removed

Has this been discussed yet at all? Typically we put public API removals through a deprecation cycle rather than executing them immediately.

broesler commented:
> > The only external API change in SciPy is that the `sparse.linalg.use_solver` function has been removed
> 
> Has this been discussed yet at all? Typically we put public API removals through a deprecation cycle rather than executing them immediately.

This deprecation has not been discussed, but I'd be happy to make the changes to deprecate it. Is there specific developer documentation on the proper way to do that? Or a recent example to which I can refer?

lucascolley commented:
> > > The only external API change in SciPy is that the `sparse.linalg.use_solver` function has been removed
> 
> > 
> 
> > Has this been discussed yet at all? Typically we put public API removals through a deprecation cycle rather than executing them immediately.
> 
> 
> 
> This deprecation has not been discussed, but I'd be happy to make the changes to deprecate it. Is there specific developer documentation on the proper way to do that? Or a recent example to which I can refer?

For very simple cases, a proposal like https://discuss.scientific-python.org/t/deprecate-stats-tiecorrect/2438/1 suffices. For more complex/impactful cases, we're looking for a proposal more akin to this: https://discuss.scientific-python.org/t/dep-depcrecate-lti-classes-in-signal-module/2432/3.

This is definitely closer to the former case, but it would be good to explain how you assess the impact of the removal and what the transition guidance would be to users. If the transition is easy enough, we may be happy to make a hard break rather than going through a deprecation cycle, but that needs discussion.

More info at https://scipy.github.io/devdocs/dev/core-dev/index.html#deprecations.
